src/transform/src/fusion/reduce.rs

```diff
@@ -0,0 +1,105 @@
+// by the Apache License, Version 2.0.
+
+//! Fuses reduce operators with parent operators if possible.
+
+use crate::TransformArgs;
+use expr::{AggregateExpr, AggregateFunc, MirRelationExpr, MirScalarExpr};
+
+/// Fuses reduce operators with parent operators if possible.
+#[derive(Debug)]
+pub struct Reduce;
+
+impl crate::Transform for Reduce {
+    fn transform(
+        &self,
+        relation: &mut MirRelationExpr,
+        _: TransformArgs,
+    ) -> Result<(), crate::TransformError> {
+        relation.visit_mut_pre(&mut |e| {
+            self.action(e);
+        });
+        Ok(())
+    }
+}
+
+impl Reduce {
+    /// Fuses reduce operators with parent operators if possible.
+    pub fn action(&self, relation: &mut MirRelationExpr) {
+        if let MirRelationExpr::Reduce {
+            input,
+            group_key,
+            aggregates,
+            monotonic: _,
+            expected_group_size: _,
+        } = relation
+        {
+            if let MirRelationExpr::Reduce {
+                input: inner_input,
+                group_key: inner_group_key,
+                aggregates: inner_aggregates,
+                monotonic: _,
+                expected_group_size: _,
+            } = &mut **input
+            {
+                // Do nothing if outer keys are not a subset of inner keys
+                if !group_key.iter().all(|k| inner_group_key.contains(k)) {
+                    return;
+                }
+
+                if aggregates.is_empty() && inner_aggregates.is_empty() {
+                    // Drop redundant inner reduce
+                    *input = Box::new(inner_input.take_dangerous());
+                } else if aggregates.len() == 1 && inner_aggregates.len() == 1 {
+                    // TODO: This can be more general than just len() == 1
+                    // Drop inner reduce and rewrite outer reduce if possible
+                    if let AggregateExpr {
+                        func: AggregateFunc::Count,
+                        expr,
+                        distinct: _,
+                    } = inner_aggregates.first().unwrap()
+                    {
+                        if inner_group_key.contains(expr) {
+                            return;
+                        }
+
+                        if let AggregateExpr {
+                            func: AggregateFunc::SumInt64,
+                            expr: MirScalarExpr::Column(i),
+                            distinct: _,
+                        } = aggregates.first().unwrap()
+                        {
+                            if *i != inner_group_key.len() {
+                                return;
+                            }
+
+                            aggregates[0] = inner_aggregates[0].clone();
+                            *input = Box::new(inner_input.take_dangerous());
+
+                            let mut outputs = Vec::with_capacity(group_key.len());
```

Comment:
Could also write this as something like

```rust
let mut output = (0..group_key.len()).collect::<Vec<usize>>();
p.push(group_key.len() + 1);
```

if preferred.

Reply:
Limited scope of PR to cases where the outer reduction doesn't depend at all on the result of any aggregation performed by the inner reduction, so this is not relevant.